crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,153 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
     duration.as_nanos().try_into().map_err(Into::into)
 }
 
+fn stdin_nonempty() -> bool {
+    use std::io::Read;
+    std::io::stdin().bytes().peekable().peek().is_some()
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use crate::fdentry::Descriptor;
+    if fd_events.is_empty() && timeout.is_none() {
+        return Ok(());
+    }
+
+    // Currently WASI file support is only (a) regular files (b) directories (c) symlinks on Windows,
+    // which are always ready to write on Unix.
+    //
+    // We need to consider stdin/stdout/stderr separately. We treat stdout/stderr as always ready to write
+    // and only poll the stdin.
+    let mut stdin_events = vec![];
+    let mut immediate_events = vec![];
+
+    for event in fd_events {
+        match event.descriptor {
+            Descriptor::Stdin if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                stdin_events.push(event)
+            }
+            _ => immediate_events.push(event),
+        }
+    }
+
+    // we have at least one immediate event, so we don't need to care about stdin
+    if immediate_events.len() > 0 {
+        for event in immediate_events {
+            let size = match event.descriptor {
+                Descriptor::OsFile(os_file) if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                    os_file
+                        .metadata()
+                        .expect("FIXME return a proper error")
+                        .len()
```

Comment:
I don't think we'll want to assume the entire file can be read regardless of how we end up implementing this.  We're obviously limited to what Windows provides us, but I think "you can read *all* of the file now without blocking" is not how we want to answer a read poll.

Reply:
See this relevant issue: https://github.com/bytecodealliance/wasmtime/issues/578

Reply:
If this is how `fionread` behaves for a file handle, then I agree with should replicate it here.

My comment was based upon my misunderstanding that we would want to do asynchronous I/O on files via this API.

I think this comment can be resolved.